doc/source/train/examples/intel_gaudi/llama.ipynb

```diff
@@ -0,0 +1,831 @@
+    "    if execution_mode == \"lazy\":\n",
+    "        os.environ[\"PT_HPU_LAZY_MODE\"] = \"1\"\n",
+    "    else:\n",
+    "        os.environ[\"PT_HPU_LAZY_MODE\"] = \"0\"\n",
+    "\n",
+    "    # Preparing train configurations\n",
+    "    train_config = {\n",
+    "        \"execution_mode\": execution_mode,\n",
+    "        \"model\": \"/root/models/llama-7b\",\n",
+    "        \"model_config\": {\"torch_dtype\": torch.bfloat16, \"trust_remote_code\": False, \"use_auth_token\": None},\n",
+    "        \"lora_config\": {\"task_type\": \"CAUSAL_LM\", \"r\": 8, \"lora_alpha\": 32, \"lora_dropout\": 0.1, \"target_modules\": [\"q_proj\", \"v_proj\"]},\n",
+    "        \"lr\": 1e-4,\n",
+    "        \"epochs\": 2,\n",
+    "        \"batch_size_per_worker\": 8,\n",
+    "        \"output\": \"/tmp/ray/\",\n",
+    "    }\n",
+    "\n",
+    "    # Configure computation resources\n",
+    "    # In ScalingConfig, require an HPU for each worker\n",
+    "    scaling_config = ScalingConfig(num_workers=num_workers, resources_per_worker={\"CPU\": 1, \"HPU\": 1})\n",
+    "    # Set backend to hccl in TorchConfig\n",
+    "    torch_config = TorchConfig(backend = \"hccl\")\n",
+    "\n",
+    "    # start your ray cluster\n",
+    "    ray.init()\n",
+    "\n",
+    "    # Initialize a Ray TorchTrainer\n",
+    "    trainer = TorchTrainer(\n",
+    "        train_loop_per_worker=train_func_per_worker,\n",
+    "        train_loop_config=train_config,\n",
+    "        torch_config=torch_config,\n",
+    "        scaling_config=scaling_config,\n",
+    "    )\n",
+    "\n",
+    "    result = trainer.fit()\n",
+    "    print(f\"Training result: {result}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Start Training\n",
+    "\n",
+    "Finally, we call the `train_llama` function to start the training process. You can adjust the number of workers to use, and the execution mode for HPU."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "<div class=\"tuneStatus\">\n",
+       "  <div style=\"display: flex;flex-direction: row\">\n",
+       "    <div style=\"display: flex;flex-direction: column;\">\n",
+       "      <h3>Tune Status</h3>\n",
+       "      <table>\n",
+       "<tbody>\n",
+       "<tr><td>Current time:</td><td>2024-04-18 07:34:56</td></tr>\n",
+       "<tr><td>Running for: </td><td>00:02:46.83        </td></tr>\n",
+       "<tr><td>Memory:      </td><td>95.6/1007.4 GiB    </td></tr>\n",
+       "</tbody>\n",
+       "</table>\n",
+       "    </div>\n",
+       "    <div class=\"vDivider\"></div>\n",
+       "    <div class=\"systemInfo\">\n",
+       "      <h3>System Info</h3>\n",
+       "      Using FIFO scheduling algorithm.<br>Logical resource usage: 5.0/160 CPUs, 0/0 GPUs (0.0/1.0 TPU, 4.0/8.0 HPU)\n",
```

Comment:
<img width="792" alt="Screenshot 2024-05-10 at 1 16 30 PM" src="https://github.com/ray-project/ray/assets/26745457/b9ce44ee-c9a4-4eb1-9648-b83d7313a23f">

Nit: saw there is a TPU in your ray cluster. Is it colocated with HPUs on the same host? Or it's just another TPU node.

Reply:
Maybe this is ray's device detection bug I think.

Reply:
Got it. Let me report this bug to Ray Core team.